mount/mount_darwin.go

```diff
@@ -0,0 +1,138 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package mount
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+
+	"github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/log"
+)
+
+var (
+	// ErrNotImplementOnUnix is returned for methods that are not implemented
+	ErrNotImplementOnUnix = errors.New("not implemented under darwin")
+	// allowedHelperBinaries is a list of third-party executable, which can
+	// be extended in future uses, e.g., mount_nfs, mount_9p.  `mount_darwin`
+	// is a special case where we will use the internal darwin snapshotter.
+	allowedHelperBinaries = []string{"mount_containerd_darwin"}
+)
+
+// Mount to the provided target path.
+//
+func (m *Mount) Mount(target string) error {
+	log.L.Debugf("mount: %s, src %s, type %s", target, m.Source, m.Type)
+
+	binary := fmt.Sprintf("mount_containerd_%s", m.Type)
+	if _, err := exec.LookPath(binary); err != nil {
+		return fmt.Errorf("mount binary not found: '%s(%s)': %w", m.Type, binary, errdefs.ErrNotFound)
+	}
+
+	for _, helperBinary := range allowedHelperBinaries {
+		if binary == helperBinary {
+			return mountWithHelper(binary, target, m.Source, m.Options)
+		}
+	}
+
+	return fmt.Errorf("mount binary isn't allowed to use: '%s(%s)': %w", m.Type, binary, ErrNotImplementOnUnix)
+}
+
+// Unmount the provided mount path with the flags
+func Unmount(target string, flags int) error {
+	return unmount(target, flags)
+}
+
+// UnmountAll the provided mount path with the flags
+func UnmountAll(target string, flags int) error {
+	return unmount(target, flags)
+}
+
+func unmount(target string, flags int) error {
+	log.L.Debugf("unmount: %s", target)
+
+	if !isMounted(target) {
+		log.L.Debugf("%s is already unmounted. skipped.", target)
+		return nil
+	}
+
+	return unmountWithHelper("mount_containerd_darwin", target, flags)
```

Comment:
Just passing by. Seems a bit odd that we use type-specific binary for mount, but a generic (?) "darwin" binary for unmount. What if my mount_containerd_foobar helper does something weird that's not actually mounting anything, in which case mount_containerd_darwin wouldn't know know to unmount it? I see that the type information is not available here but wanted to ask anyway, in case it's possible to pass it in and if you agree it makes sense.

Reply:
We saw something like this in the Windows mount logic (type information not available in unmount to distinguish one of several mounting mechanisms, and the proposed solution used [there](https://github.com/containerd/containerd/pull/4419/files#diff-ede639814a8abed289b5808723cc84b4aeee754dfc97fbaa435fb0dc4e910b7d) is to stash the information in the filesystem. So I agree that the unmounter should be the same as the mounter to support use-cases such as this.